src/locale/keymap-util.c

```diff
@@ -780,3 +782,228 @@ int x11_convert_to_vconsole(Context *c) {
+                char *line_locale;
+
+                r = read_line(fr, LONG_LINE_MAX, &line);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                if (locale_enabled) {
+                        /* Just complete writing the file if the new locale was already enabled */
+                        if (!first_line)
+                                fputc('\n', fw);
+                        fputs(line, fw);
+                        first_line = false;
+                        continue;
+                }
+
+                line = strstrip(line);
+                if (isempty(line)) {
+                        fputc('\n', fw);
+                        first_line = false;
+                        continue;
+                }
+
+                line_locale = line;
+                if (line_locale[0] == '#')
+                        line_locale = strstrip(line_locale + 1);
+                else if (strcasecmp_ptr(line_locale, locale_entry) == 0)
+                        return 0; /* the file already had our locale activated, so skip updating it */
+
+                if (strcasecmp_ptr(line_locale, locale_entry) == 0) {
+                        /* Uncomment existing line for new locale */
+                        if (!first_line)
+                                fputc('\n', fw);
+                        fputs(locale_entry, fw);
+                        locale_enabled = true;
+                        first_line = false;
+                        continue;
+                }
+
+                /* The line was not for the locale we want to enable, just copy it */
+                if (!first_line)
+                        fputc('\n', fw);
+                fputs(line, fw);
+                first_line = false;
+        }
+
+        /* Add locale to enable to the end of the file if it was not found as commented line */
+        if (!locale_enabled)
+                fprintf(fw, "\n%s", locale_entry);
+        fputc('\n', fw);
+
+        r = fflush_sync_and_check(fw);
+        if (r < 0)
+                return r;
+
+        if (rename(temp_path, "/etc/locale.gen") < 0)
+                return -errno;
+        temp_path = mfree(temp_path);
+
+        return 0;
+#else
+        return -EOPNOTSUPP;
+#endif
+}
+
+int locale_gen_run(void) {
+#if HAVE_LOCALEGEN
+        pid_t pid;
+        int r;
+
+        /* Silently do nothing if locale-gen was not available */
+        if (!locale_gen_check_available())
+                return 0;
+
+        r = safe_fork("(sd-localegen)", FORK_RESET_SIGNALS|FORK_RLIMIT_NOFILE_SAFE|FORK_CLOSE_ALL_FDS|FORK_DEATHSIG|FORK_LOG|FORK_WAIT, &pid);
+        if (r < 0)
+                return r;
+        if (r == 0) {
+                execl(LOCALEGEN_PATH, LOCALEGEN_PATH, NULL);
+                log_error_errno(errno, "Failed to invoke locale-gen: %m");
+                _exit(EXIT_FAILURE);
+        }
+
+        return 0;
+#else
+        return -EOPNOTSUPP;
```

Comment:
This seems a bit inconsistent. 0 is returned if HAVE_LOCALEGEN was configured but not found, and -EOPNOTSUPP if it wasn't configured.

Reply:
Yes - quite honestly, this should fail properly if locale-gen wasn't found, because *if* this function is called, we expect it to run locale-gen or return `-EOPNOTSUPP` if support for it wasn't compiled in.